File: cpp/tests/test_deconvolution.cpp
/*
 * Hydrophone Deconvolution - Multi-language Implementation
 *
 * Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
 * DOI: 10.5281/zenodo.10079801
 * Original License: CC BY 4.0
 *
 * This implementation: 2024
 * License: CC BY 4.0
 */

#include "deconvolution.hpp"
#include <iostream>
#include <cmath>
#include <random>
#include <unsupported/Eigen/FFT>

int main() {
    std::cout << "=== C++ Deconvolution Tests ===" << std::endl;
    int passed = 0, failed = 0;

    // Test 1: Output length
    std::cout << "Test 1: Output length... ";
    {
        int n = 1024;
        std::mt19937 gen(42);
        std::normal_distribution<> dist(0.0, 1.0);

        Eigen::VectorXd signal(n);
        Eigen::VectorXd temp(n);
        for (int i = 0; i < n; ++i) {
            signal(i) = dist(gen);
            temp(i) = dist(gen);
        }

        Eigen::FFT<double> fft;
        Eigen::VectorXcd freq_resp;
        fft.fwd(freq_resp, temp);
        for (int i = 0; i < n; ++i) freq_resp(i) += 1.0;

        auto result = hydrophone::deconvolve_without_uncertainty(signal, freq_resp, 1e7);
        if (result.size() == n) {
            std::cout << "PASSED" << std::endl;
            passed++;
        } else {
            std::cout << "FAILED" << std::endl;
            failed++;
        }
    }

    // Test 2: Known signal recovery
    std::cout << "Test 2: Known signal recovery... ";
    {
        int n = 256;
        std::mt19937 gen1(42);
        std::normal_distribution<> dist(0.0, 1.0);

        Eigen::VectorXd original(n);
        for (int i = 0; i < n; ++i) original(i) = dist(gen1);

        std::mt19937 gen2(99);
        Eigen::VectorXd temp(n);
        for (int i = 0; i < n; ++i) temp(i) = dist(gen2);

        Eigen::FFT<double> fft;
        Eigen::VectorXcd freq_resp;
        fft.fwd(freq_resp, temp);
        for (int i = 0; i < n; ++i) freq_resp(i) += 2.0;

        Eigen::VectorXcd original_fft;
        fft.fwd(original_fft, original);

        Eigen::VectorXcd convolved_fft(n);
        for (int i = 0; i < n; ++i) {
            convolved_fft(i) = original_fft(i) * freq_resp(i);
        }

        Eigen::VectorXd measured;
        fft.inv(measured, convolved_fft);

        auto recovered = hydrophone::deconvolve_without_uncertainty(measured, freq_resp, 1e7);

        double max_err = (recovered - original).cwiseAbs().maxCoeff();
        if (max_err < 1e-10) {
            std::cout << "PASSED (max error: " << max_err << ")" << std::endl;
            passed++;
        } else {
            std::cout << "FAILED (max error: " << max_err << ")" << std::endl;
            failed++;
        }
    }

    std::cout << "\n=== Results: " << passed << " passed, " << failed << " failed ===" << std::endl;
    return failed > 0 ? 1 : 0;
}

File: rust/tests/test_deconvolution.rs
//! Hydrophone Deconvolution - Multi-language Implementation
//!
//! Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
//! DOI: 10.5281/zenodo.10079801
//! License: CC BY 4.0

use hydrophone_deconvolution::deconvolve_without_uncertainty;
use num_complex::Complex64;
use rustfft::FftPlanner;

#[test]
fn test_output_length() {
    let n = 1024;
    let signal: Vec<f64> = (0..n).map(|i| (i as f64 * 0.1).sin()).collect();
    let freq_resp: Vec<Complex64> = (0..n).map(|_| Complex64::new(1.0, 0.0)).collect();
    let result = deconvolve_without_uncertainty(&signal, &freq_resp, 1e7);
    assert_eq!(result.len(), n);
}

#[test]
fn test_known_signal_recovery() {
    let n = 256;
    let original: Vec<f64> = (0..n).map(|i| (i as f64 * 0.05).sin()).collect();

    // Create frequency response via FFT (well-conditioned) + offset
    let impulse: Vec<f64> = (0..n).map(|i| if i == 0 { 2.0 } else { 0.1 * (i as f64 * 0.01).sin() }).collect();
    let mut freq_resp: Vec<Complex64> = impulse.iter().map(|&x| Complex64::new(x, 0.0)).collect();
    let mut planner = FftPlanner::new();
    let fft = planner.plan_fft_forward(n);
    fft.process(&mut freq_resp);

    // Convolve: FFT(original) * freq_resp, IFFT, scale
    let mut original_fft: Vec<Complex64> = original.iter().map(|&x| Complex64::new(x, 0.0)).collect();
    fft.process(&mut original_fft);

    let mut convolved: Vec<Complex64> = original_fft.iter().zip(freq_resp.iter())
        .map(|(s, f)| s * f).collect();

    let ifft = planner.plan_fft_inverse(n);
    ifft.process(&mut convolved);

    let scale = 1.0 / n as f64;
    let measured: Vec<f64> = convolved.iter().map(|c| c.re * scale).collect();

    let recovered = deconvolve_without_uncertainty(&measured, &freq_resp, 1e7);

    let max_err = recovered.iter().zip(original.iter())
        .map(|(r, o)| (r - o).abs())
        .fold(0.0f64, f64::max);

    assert!(max_err < 1e-10, "Max error: {}", max_err);
}
